File: leetcode/1001-2000/1801-1900/1881-1890/1888-minimum-number-of-flips-to-make-the-binary-string-alternating/MinimumNumberOfFlipsToMakeTheBinaryStringAlternating.java
public class MinimumNumberOfFlipsToMakeTheBinaryStringAlternating {

  public static void main(String[] args) {
    var solution = new Solution();

    System.out.println(solution.minFlips("111000"));
    System.out.println(solution.minFlips("1110"));
  }
}

class Solution {
  public int minFlips(String s) {
    var operations = new int[] { 0, 0 };

    var minimum = Integer.MAX_VALUE;

    for (var i = 0; i < s.length(); i++) {
      operations[(i ^ s.charAt(i)) & 1]++;
    }

    minimum = Math.min(operations[0], operations[1]);

    for (var i = 0; i < s.length(); i++) {
      operations[(i ^ s.charAt(i)) & 1]--;
      operations[((i + s.length()) ^ s.charAt(i)) & 1]++;

      minimum = Math.min(Math.min(operations[0], operations[1]), minimum);
    }

    return minimum;
  }
}

File: leetcode/1001-2000/1801-1900/1881-1890/1888-minimum-number-of-flips-to-make-the-binary-string-alternating/MinimumNumberOfFlipsTomakeTheBinaryStringAlternating.ts
function minFlips(s: string): number {
  let minimum = Number.MAX_SAFE_INTEGER;
  const minimums = [0, 0];

  for (let i = 0; i < s.length; i++) {
    minimums[(i ^ s.charCodeAt(i)) & 1]++;
  }

  for (let i = 0; i < s.length; i++) {
    minimums[(i ^ s.charCodeAt(i)) & 1]--;
    minimums[((i + s.length) ^ s.charCodeAt(i)) & 1]++;

    minimum = Math.min(...minimums, minimum);
  }

  return minimum;
}

console.log(minFlips("01001001101"));
// console.log(minFlips("111000"));
// console.log(minFlips("1110"));
